### PR TITLE
Add documentation for AD module

### LINKED ISSUE #6172: Improve automatic differentiation documentation
@jppelteret Would it be possible to improve the documentation [here](http://dealii.org/developer/doxygen/deal.II/group__auto__symb__diff.html). There are a couple of `Todo` left.

### PR DESCRIPTION
**Pull request by jppelteret**
Fixes #6172

### CODE REVIEW

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 1]
```diff
@@ -14,47 +14,389 @@
...
+ * within deal.II's framework.
  *
  * @section auto_diff_1 Automatic differentiation
  *
- * @todo Write a short introduction into AD. As a temporary entry, the following links
- * may be enlightening:
- * - <a href="https://en.wikipedia.org/wiki/Automatic_differentiation">Wikipedia article</a>
- * - <a href="https://projects.coin-or.org/ADOL-C/browser/stable/2.6/ADOL-C/doc/adolc-manual.pdf?format=raw#page=1>Adol-C manual</a>
+ * Automatic differentiation (commonly also referred to as called algorithmic differentiation),
+ * is a that can be used to "automatically" compute the first, and perhaps higher-order,
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: is a that

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 2]
```diff
@@ -14,47 +14,389 @@
...
+ * derivatives of function(s) with respect to one or more input variables.
+ * Although this comes at a certain computational cost, the benefits to using such a tool may be
+ * significant. When used correctly the derivatives of often complicated functions can be computed
+ * to a very high accuracy. Although the exact accuracy achievable by these frameworks largely
+ * depends their underlying mathematical formulation, some implementations compute with a precision
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: depends on ?

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 3]
```diff
@@ -14,47 +14,389 @@
...
+ * computations, lies somewhere between these two limiting cases. Its computation for a general composite
+ * function remains an open problem in graph theory.
+ *
+ * With the aid of the diagram below (it and some of the listed details courtesy of this
+ * <a href="https://en.wikipedia.org/wiki/Automatic_differentiation">Wikipedia article</a>),
+ *
+ * <div class="twocolumn" style="width: 80%">
+ *   <div class="parent">
+ *     <div class="img" align="center">
+ *       <img src="https://upload.wikimedia.org/wikipedia/commons/a/a4/ForwardAccumulationAutomaticDifferentiation.png"
```
Comment (jppelteret) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: These two images
```
https://commons.wikimedia.org/wiki/File:ForwardAccumulationAutomaticDifferentiation.png
https://commons.wikimedia.org/wiki/File:ReverseaccumulationAD.png
```
have been released into the public domain. Does this then mean that we can host copies ourselves that can be downloaded with the offline documentation?

**Comment by jppelteret** [comment]
This is my first pass at writing this documentation. If I have time tomorrow afternoon I will go through it again. However, any comments and feedback between now and then would be appreciated.

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 4]
```diff
@@ -14,47 +14,389 @@
...
+ * one of three categories. In each, some customized data classes representing the floating point value
+ * of an evaluated function and its derivative(s) by
+ * -# exploiting <em>dual</em> / <em>complex-step</em> / <em>hyper-dual</em> formulations (occasionally
+ *    called <em>tapeless</em> methods),
+ * -# those utilizing <em>taping</em> strategies, and
+ * -# those using compile-time optimization through <em>expression templates</em>.
+ *
+ * To provide some tentative insight into how these various implementations might look like in practice, we
+ * offer the following generic summary of these approaches:
+ * -# The first two <em>tapeless</em> approaches listed above use some variation of a truncated Taylor
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: you list three and the second is `taping`, so which of those are meant here?

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 5]
```diff
@@ -14,47 +14,389 @@
...
+ *    series, along with a particular choice of definition for the perturbation parameter, to compute
+ *    function derivatives using a finite-difference based approach. The "dual" number constitutes the
+ *    accumulated directional derivatives computed simultaneously as the function values is evaluated; in the
+ *    complex-step approach, the imaginary value effectively serves this purpose. The choice of the perturbation
+ *    parameter determines the numerical qualities of the scheme, such as the influence of the truncation of
+ *    the Taylor scheme; dual numbers do not contain any higher-order terms in their first derivative, while
+ *    the for the complex-step method these existent higher-order terms are neglected. It can be shown that
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: the for the

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 6]
```diff
@@ -14,47 +14,389 @@
...
+ *    numerical perturbation. The dual number approach thus produces exact first derivatives, while the
+ *    complex-step approach does not. The standard implementation of the dual numbers, however, cannot yield
+ *    exact values for second derivatives. Hyper-dual numbers take a different view of this idea, with the
+ *    outcome that both first and second derivatives can be computed exactly.
+ * -# With <em>taped</em> approaches, a specified subregion of code is selected as one for which all
+ *    operations executed with active (marked) input variables are tracked and recorded in a data structure
+ *    referred to as a tape. At the end of the taped region, the recorded function(s) may be revaluated
+ *    by "replaying" the tape with a different set of input variables instead of recomputing the function
+ *    directly. Assuming that the taped region represents a smooth function, arbitrarily high-order
+ *    derivatives of the function then can be by referring to the code path computed and stored on the tape.
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: can be by referring

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 7]
```diff
@@ -14,47 +14,389 @@
...
+ *    (This could perhaps be achieve, for example, through evaluation of the function around the point
+ *    of interest.) There exist strategies to deal with situations where the taped function is not
+ *    smooth at the evaluated point, or if it is not analytic. Furthermore, one might need to consider the
+ *    case of branched functions, where the tape is no longer sequential, but rather forks off on a different
+ *    evaluation path to that due to the original recorded inputs.
+ * -# Methods based on <a href="https://en.wikipedia.org/wiki/Expression_templates">expression templates</a>
+ *    leverage the computational graph
+ *    (in this case, a <a href="https://en.wikipedia.org/wiki/Directed_acyclic_graph">directed acyclic graph (DAG)</a>),
+ *    constructed from the abstract syntax tree (AST), that resolves the function output from its input values.
+ *    The outermost leaves on the represent the independent variables or constants, and are transformed by unary
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: leaves ~~on the~~ represent ?

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 8]
```diff
@@ -14,47 +14,389 @@
...
+ *
+ * representing the calculation of the function
+ * \f[
+ *   f (\mathbf{x})
+ *     = x_{1} \times x_{2} + \sin (x_{1})
+ *   \quad ,
+ * \f]
+ * we will briefly describe what forward- and reverse- auto-differentiation are.
+ * Note that in the diagram, along the edges of the graph in text are the directional
+ * derivative of function \f$ w \f$ with respect to the i-th variable, represented by
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: here and below it should be enough to do `$`, AFAIR dealii does some preprocessing. But check compiled page to be sure

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 9]
```diff
@@ -14,47 +14,389 @@
...
  *
  * Since the interface to each automatic differentiation library is so vastly different,
- * a uniform internal interface to each number has been established. This allows the
- * driver classes (that provide the core functionality, and are introduced in the next
- * section) a consistent mechanism to interact with different auto-differentiation
- * libraries. Specifically, they need to be able to correctly initialize and finalize data
- * that is to be interpreted as the dependent and independent variables of a formula.
+ * a uniform internal interface to each number will been established in the near future.
+ * The goal will be to allow some driver classes (that provide the core functionality,
+ * and will later be introduced in the next section) a consistent mechanism to interact with
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: have a consistent ? (verb is missing)

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h (hunk 4 above)
Comment (jppelteret) [reply doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: No, this was referring to "dual" and "complex-step" numbers (with the third being "hyper-dual"). I've clarified this in the text now.

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h (hunk 8 above)
Comment (jppelteret) [reply doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: Yes, you are right. I will change this!

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h (hunk 3 above)
Comment (bangerth) [reply doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: I would just keep the link and not bother with hosting it ourselves.

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 10]
```diff
@@ -14,47 +14,395 @@
...
+ * within deal.II's framework.
  *
  * @section auto_diff_1 Automatic differentiation
  *
- * @todo Write a short introduction into AD. As a temporary entry, the following links
- * may be enlightening:
- * - <a href="https://en.wikipedia.org/wiki/Automatic_differentiation">Wikipedia article</a>
- * - <a href="https://projects.coin-or.org/ADOL-C/browser/stable/2.6/ADOL-C/doc/adolc-manual.pdf?format=raw#page=1>Adol-C manual</a>
+ * <a href="https://en.wikipedia.org/wiki/Automatic_differentiation">Automatic differentiation </a>
+ * (commonly also referred to as called algorithmic differentiation),
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: ~~called~~

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 11]
```diff
@@ -14,47 +14,395 @@
...
+ *   title     = {The Development of Hyper-Dual Numbers for Exact Second-Derivative Calculations},
+ *   booktitle = {49th {AIAA} Aerospace Sciences Meeting including the New Horizons Forum and Aerospace Exposition},
+ *   year      = {2011},
+ *   volume    = {886},
+ *   pages     = {124},
+ *   month     = {jan},
+ *   publisher = {American Institute of Aeronautics and Astronautics},
+ *   doi       = {10.2514/6.2011-886},
+ * }
+ * @endcode
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: you can keep the two inside a single `@code{.bib}`

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 12]
```diff
@@ -14,47 +14,395 @@
...
+ *   title     = {Algorithm 755: {ADOL-C}: a package for the automatic differentiation of algorithms written in {C/C++}},
+ *   journal   = {ACM Transactions on Mathematical Software (TOMS)},
+ *   year      = {1996},
+ *   volume    = {22},
+ *   number    = {2},
+ *   pages     = {131--167},
+ *   doi       = {10.1145/229473.229474},
+ *   publisher = {ACM}
+ * }
+ * @endcode
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: same here, you can group those into a single `@code{.bib}`

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 13]
```diff
@@ -14,47 +14,395 @@
...
+ *   title         = {Automatic Differentiation of C++ Codes for Large-Scale Scientific Computing},
+ *   booktitle     = {International Conference on Computational Science {\textendash} {ICCS} 2006},
+ *   publisher     = {Springer Berlin Heidelberg},
+ *   year          = {2006},
+ *   editor        = {Alexandrov, V.N. and van Albada, G.D. and Sloot, P.M.A. amd Dongarra, J.},
+ *   pages         = {525--532},
+ *   doi           = {10.1007/11758549_73},
+ *   organization  = {Springer}
+ * }
+ * @endcode
```
Comment (davydden) [inline doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: and here

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h (hunk 11 above)
Comment (jppelteret) [reply doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: I showed @davydden that this currently renders each reference within its own "code block". I prefer this style, and he's agreed that I need not change it.

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 14]
```diff
@@ -14,47 +14,395 @@
...
+ * Consider first that any composite function $f(x)$, here represented as having two
+ * independent variables, can be dissected into a composition of its elementary functions
+ * @f[
+ *   f (\mathbf{x})
+ *   = f_{0} \circ f_{1} \circ f_{2} \circ \ldots \circ f_{n} (\mathbf{x})
+ *   \quad .
+ * @f]
+ * As was previously mentioned, if each of the primitive operations $f_{n}$ is smooth and
+ * differentiable, then the chain can be universally employed to compute the total derivative of $f$,
+ * namely $\dfrac{d f(x)}{d \mathbf{x}}$. How exactly the chain-rule is applied is what
```
[the remark this answers is not in the document]
Comment (jppelteret) [reply doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: Yes, I can see how this makes for awkward reading. I'm going to rephrase it if you don't mind.

File: doc/doxygen/headers/automatic_and_symbolic_differentiation.h [hunk 15]
```diff
@@ -14,47 +14,395 @@
...
+ *   doi       = {10.1.1.210.4834},
+ *   pages     = {181--202},
+ *   url       = {https://projects.coin-or.org/ADOL-C/browser/trunk/ADOL-C/doc/adolc-manual.pdf}
+ * }
+ * @endcode
+ *
+ * provides the principle insights into their taped and tapeless implementations, and how ADOL-C
+ * can be incorporated into a user code.
+ * Some further useful resources for understanding the implementation of ADOL-C, and possibilities
+ * for how it may be used within a numerical code, include
```
[the remark this answers is not in the document]
Comment (jppelteret) [reply doc/doxygen/headers/automatic_and_symbolic_differentiation.h]: Here and below there's no elegant way to finish the sentence with a period. I'd have to start the next paragraph with one, and that's not great either. So I'm going with some alternative that suggests that each bib entry is part of a list. If you don't like it, just say so.

**Comment by jppelteret** [comment]
Thank you very much for the feedback @masterleinad!

**Pull request merged**